ln-gateway/src/lib.rs

```diff
@@ -1,75 +1,67 @@
 pub mod ln;
+pub mod webserver;
 
 use crate::ln::{LightningError, LnRpc};
 use bitcoin_hashes::sha256::Hash;
 use minimint::modules::ln::contracts::{incoming::Preimage, ContractId};
-use minimint_api::{db::Database, Amount, OutPoint, TransactionId};
-use mint_client::clients::gateway::{GatewayClient, GatewayClientConfig, GatewayClientError};
+use minimint_api::{Amount, OutPoint};
+use mint_client::clients::gateway::{GatewayClient, GatewayClientError};
 use rand::{CryptoRng, RngCore};
-use serde::{Deserialize, Serialize};
-use std::path::PathBuf;
-use std::sync::Arc;
-use std::time::Duration;
-use std::time::Instant;
+use std::{
+    sync::Arc,
+    time::{Duration, Instant},
+};
 use thiserror::Error;
 use tokio::sync::Mutex;
 use tracing::{debug, error, instrument, warn};
+use webserver::run_webserver;
 
-#[derive(Debug, Serialize, Deserialize)]
-pub struct LnGatewayConfig {
-    pub federation_client: GatewayClientConfig,
-    pub ln_socket: PathBuf,
-}
+pub type SharedLnGateway = Arc<Mutex<Option<LnGateway>>>;
 
 pub struct LnGateway {
     federation_client: Arc<GatewayClient>,
     ln_client: Arc<dyn LnRpc>,
     fetcher: tokio::task::JoinHandle<()>,
+    // TODO: what does this tide::Result mean?
+    webserver: tokio::task::JoinHandle<tide::Result<()>>,
 }
 
 impl LnGateway {
-    pub async fn from_config(db: Box<dyn Database>, cfg: LnGatewayConfig) -> LnGateway {
-        let LnGatewayConfig {
-            federation_client,
-            ln_socket,
-        } = cfg;
-        let federation_client = GatewayClient::new(federation_client, db);
-        let ln_client = cln_rpc::ClnRpc::new(ln_socket)
-            .await
-            .expect("connect to ln_socket");
-        let ln_client = Mutex::new(ln_client);
-
-        Self::new(Arc::new(federation_client), Box::new(ln_client)).await
-    }
-
-    pub async fn new(
+    pub async fn run(
```

Comment:
Not sure what to do with this API. Seems moving background fetch handle into this struct is good. And we need to pass the mutex to the background fetcher and webserver to avoid concurrency problems. But it's a little weird to pass it as argument here ...

Also just realized I still need to pass the SharedLnGateway to background_fetch ...